pkg/backends/calico/client.go

```diff
@@ -188,14 +235,241 @@ func (c *client) OnStatusUpdated(status api.SyncStatus) {
 	}
 }
 
+type BGPPeer struct {
+	PeerIP      net.IP               `json:"ip"`
+	ASNum       numorstring.ASNumber `json:"as_num,string"`
+	RRClusterID string               `json:"rr_cluster_id"`
+}
+
+func (c *client) updatePeersV1() {
+	// A map that will contain the v1 peerings that should exist, with the same key and
+	// value form as c.peeringCache.
+	peersV1 := make(map[string]string)
+
+	// Common subroutine for emitting both global and node-specific peerings.
+	emit := func(key model.Key, peer *BGPPeer) {
+		log.WithFields(log.Fields{"key": key, "peer": peer}).Debug("Maybe emit peering")
+
+		// Compute etcd v1 path for this peering key.
+		k, err := model.KeyToDefaultPath(key)
+		if err != nil {
+			log.Errorf("Ignoring update: unable to create path from Key %v: %v", key, err)
+			return
+		}
+
+		// If we already have an entry for that path, it wins.  When we're
+		// emitting reverse peerings to ensure symmetry, this is what ensures
+		// that an explicit forwards peering is not overwritten by an implicit
+		// reverse peering.
+		if _, ok := peersV1[k]; ok {
+			log.Debug("Peering already exists")
+			return
+		}
+
+		// If we would be emitting a node-specific peering to a peer IP, and we
+		// already have a global peering to that IP, skip emitting the node-specific
+		// one.
+		if nodeKey, ok := key.(model.NodeBGPPeerKey); ok {
+			globalKey := model.GlobalBGPPeerKey{PeerIP: nodeKey.PeerIP}
+			globalPath, _ := model.KeyToDefaultPath(globalKey)
+			if _, ok = peersV1[globalPath]; ok {
+				log.Debug("Global peering already exists")
+				return
+			}
+		}
+
+		// Serialize and store the value for this peering.
+		value, err := json.Marshal(peer)
+		if err != nil {
+			log.Errorf("Ignoring update: unable to serialize value %v: %v", peer, err)
+			return
+		}
+		peersV1[k] = string(value)
+	}
+
+	// Loop through v3 BGPPeers.
+	for _, v3res := range c.bgpPeers {
+		log.WithField("peer", v3res).Debug("First pass with v3 BGPPeer")
+
+		var localNodeNames []string
+		if v3res.Spec.NodeSelector != "" {
+			localNodeNames = c.nodesMatching(v3res.Spec.NodeSelector)
+		} else if v3res.Spec.Node != "" {
+			localNodeNames = []string{v3res.Spec.Node}
+		}
+		log.Debugf("Local nodes %#v", localNodeNames)
+
+		var peers []*BGPPeer
+		if v3res.Spec.PeerSelector != "" {
+			for _, peerNodeName := range c.nodesMatching(v3res.Spec.PeerSelector) {
+				peers = append(peers, c.nodeAsBGPPeer(peerNodeName))
+			}
+		} else {
+			ip := net.ParseIP(v3res.Spec.PeerIP)
+			if ip == nil {
+				log.Warning("PeerIP is not assigned or is malformed")
+				continue
+			}
+			peers = append(peers, &BGPPeer{
+				PeerIP: *ip,
+				ASNum:  v3res.Spec.ASNumber,
+			})
+		}
+		log.Debugf("Peers %#v", peers)
+
+		for _, peer := range peers {
+			log.Debugf("Peer: %#v", peer)
+			if localNodeNames == nil {
```

Comment:
I think this ties in with the calicoctl UI question.  If we're not motivated to improve the calicoctl UI at this point, it's probably better to keep this as is, so that calicoctl can keep reporting as either "global" or "node-specific" (with selector-based peers being reported as "node-specific").

I also (as above) think there is some merit in maximal compatibility with existing templates; so overall I'd err on the side of the status quo here.

WDYT?

Reply:
Let's err on the status quo for now - I'd rather get something out sooner than nitpick over details.

But let's line up a task to reconsider this decision in the near future? I think it's probably an easy win for reducing the overall complexity of the templates and code.

Reply:
OK, I've raised a task for that.